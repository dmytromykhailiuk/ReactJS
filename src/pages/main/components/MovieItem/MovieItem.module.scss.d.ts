// This file is automatically generated.
// Please do not change this file!
interface CssExports {
  'movie': string;
  'movie__category': string;
  'movie__footer': string;
  'movie__image': string;
  'movie__image-wrapper': string;
  'movie__name': string;
  'movie__release': string;
  'movie__three-dots': string;
}
export const cssExports: CssExports;
export default cssExports;
