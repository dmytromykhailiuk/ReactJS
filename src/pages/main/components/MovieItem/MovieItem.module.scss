@import "scss/mixins.scss";

.movie {
  width: 85.7vw;
  padding-bottom: 20px;
  border-bottom: 1vw solid $medium-grey-color;
  margin-bottom: 5vw;
  cursor: pointer;

  @include small-screen {
    width: 41vw;
    border: none;
    margin-bottom: 0;
  }

  @include large-screen {
    width: 26vw;
  }

  &__image-wrapper {
    width: 100%;
    height: 132vw;
    position: relative;

    @include small-screen {
      height: 63vw;
    }

    @include large-screen {
      height: 40vw;
    }
  }

  &__image {
    width: 100%;
    height: 132vw;
    object-fit: cover;
    background-position: center 0%;
    
    @include small-screen {
      height: 63vw;
    }

    @include large-screen {
      height: 40vw;
    }
  }

  &__footer {
    margin-top: 4.5vw;
    display: flex;
    justify-content: space-between;

    @include small-screen {
      margin-top: 1.5vw;
    }

    @include large-screen {
      margin-top: 1.1vw;
    }
  }

  &__name {
    color: rgba($white-color, 0.75);
    font-size: 4vw;
    width: 75%;
    line-height: 1.3;
    letter-spacing: 0.2vw;

    @include small-screen {
      font-size: 1.8vw;
      letter-spacing: 0.1vw;
    }

    @include large-screen {
      font-size: 1.2vw;
      letter-spacing: 0.06vw;
    }
  }

  &__category {
    margin-top: 2.2vw;
    color: $light-grey-color;
    font-size: 3vw;

    @include small-screen {
      margin-top: 1vw;
      font-size: 1.6vw;
      letter-spacing: 0.03vw;
    }

    @include large-screen {
      margin-top: 0.7vw;
      font-size: 1.1vw;
      letter-spacing: 0.02vw;
    }
  }

  &__release {
    border: solid 0.4vw $light-grey-color;
    align-self: baseline;
    padding: 1vw 4vw;
    border-radius: 1vw;
    font-size: 3vw;
    color: rgba($white-color, 0.75);

    @include small-screen {
      border: solid 0.2vw $light-grey-color;
      padding: 0.6vw 2.4vw;
      border-radius: 0.4vw;
      font-size: 1.6vw;
    }

    @include large-screen {
      border: solid 0.12vw $light-grey-color;
      padding: 0.4vw 1.6vw;
      border-radius: 0.27vw;
      font-size: 1.1vw;
    }
  }

  &__three-dots {
    display: none;
    transition: 0.4s;
    position: absolute;
    background-color: $dark-grey-color;
    border-radius: 50%;
    right: 3vw;
    top: 4vw;
    width: 8vw;
    height: 8vw;
    padding: 1.3vw;

    &:hover {
      background-color: rgba($medium-grey-color, 0.7);
    }

    @include small-screen {
      right: 1.5vw;
      top: 2vw;
      width: 4vw;
      height: 4vw;
      padding: 0.65vw;
    }

    @include large-screen {
      right: 1vw;
      top: 1.3vw;
      width: 2.7vw;
      height: 2.7vw;
      padding: 0.43vw;
    }
  }

  &:hover &__three-dots {
    display: block;
  }
}
