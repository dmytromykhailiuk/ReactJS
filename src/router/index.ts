export { default as Routes } from './Routes';
export * from './history';
