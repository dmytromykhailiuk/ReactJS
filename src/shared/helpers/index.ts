export * from "./filterMovies";
export * from "./sortMovies";
export * from "./generateId";
export * from "./clickOutside";
