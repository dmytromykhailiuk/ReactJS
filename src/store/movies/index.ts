import * as MoviesAction from "./actions";
import * as MoviesSelector from "./selectors";

export { MoviesAction, MoviesSelector };
export * from "./effects";
export * from "./reducer";
