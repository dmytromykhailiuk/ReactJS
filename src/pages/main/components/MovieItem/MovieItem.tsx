import React, { useCallback, useMemo, useState } from "react";
import { Movie } from "models";
import { ThreeDotsIcon } from "shared/components";
import classes from "./MovieItem.module.scss";
import { MovieMenu } from "../";
import { isClickInside } from "shared/helpers";
import { useHistory } from "react-router";

const DEFAULT_IMG_PATH = "https://diesel.lviv.ua/img/no-picture.png";

interface MovieItemProps {
  movie: Movie;
  onEditMovie: (movie: Movie) => void;
  onDeleteMovie: (movie: Movie) => void;
}

const MovieItem: React.FC<MovieItemProps> = React.memo(({ 
  movie, 
  onEditMovie,
  onDeleteMovie,
}) => {
  const [shoudShowMenu, setShoudShowMenuValue] = useState(false);
  const [hasError, setHasErrorValue] = useState(false);

  useMemo(() => {
    setHasErrorValue(false);
  }, [movie.id]);

  const setError = useCallback(() => {
    setHasErrorValue(true);
  }, []);

  const history = useHistory();

  const onCloseMenu = useCallback(() => {
    setShoudShowMenuValue(false);
  }, []);

  const onOpenMenu = useCallback(() => {
    setShoudShowMenuValue(true);
  }, []);

  const onEditButtonClicked = useCallback(() => {
    setShoudShowMenuValue(false);
    onEditMovie(movie);
  }, [movie]);

  const onDeleteButtonClicked = useCallback(() => {
    setShoudShowMenuValue(false);
    onDeleteMovie(movie);
  }, [movie]);

  const onMovieItemClick = useCallback((event: React.SyntheticEvent) => {
    if(isClickInside(event, classes["movie__three-dots"], classes["movie__menu"])) {
      history.push(`/film/${movie.id}`);
      window.scrollTo(0, 0);
    }
  }, [])

  return (
    <div className={classes.movie} onClick={onMovieItemClick}>
      <div className={classes["movie__image-wrapper"]}>
        <img 
          src={hasError ? DEFAULT_IMG_PATH : movie.poster_path} 
          alt={movie.title}
          className={classes["movie__image"]}
          onError={setError}
        />
        { shoudShowMenu ? 
          <div 
            className={classes["movie__menu"]} 
          >
            <MovieMenu 
              uniqueClass={movie.id}
              onCloseButtonClicked={onCloseMenu}
              onEditButtonClicked={onEditButtonClicked}
              onDeleteButtonClicked={onDeleteButtonClicked}
            />
          </div> 
        : 
          <div 
            className={classes["movie__three-dots"]}
            onClick={onOpenMenu}
          >
            <ThreeDotsIcon />
          </div> 
        }
      </div>
      <div className={classes.movie__footer}>
        <div>
          <div className={classes.movie__name}>{movie.title}</div>
          <div className={classes.movie__category}>{movie.genres.join(", ")}</div>
        </div>
        <div className={classes.movie__release}>{movie.release_date.split('-')[0]}</div>
      </div>
    </div>
  );
});

export default MovieItem;
