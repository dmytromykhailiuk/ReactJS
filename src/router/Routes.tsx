import { RouterPaths } from "shared/enums";
import React from "react";
import { Redirect, Route, Switch } from "react-router";
import { ErrorPage } from "pages/error";
import { MainPage } from "pages/main";

const Routes = () => {
  return (
    <Switch>
      <Route path={RouterPaths.HOME} exact component={MainPage} />
      <Route path={RouterPaths.FILM} component={MainPage} />
      <Route path={RouterPaths.SEARCH} component={MainPage} />
      <Route path={RouterPaths.ERROR} render={({ history }) => {
        return <ErrorPage navigateToHome={() => history.push(RouterPaths.HOME)} />  
      }} />
      <Redirect to={RouterPaths.ERROR}/>
    </Switch>
  )
};

export default Routes;
