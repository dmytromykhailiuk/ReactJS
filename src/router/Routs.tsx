import { RouterPaths } from "shared/enums";
import React from "react";
import { Redirect, Route, Switch } from "react-router";
import { ErrorPage } from "pages/error";
import { MainPageContainer } from "pages/main";

const Routs = () => {
  return (
    <Switch>
      <Route path={RouterPaths.HOME} exact component={MainPageContainer} />
      <Route path={RouterPaths.FILM} component={MainPageContainer} />
      <Route path={RouterPaths.SEARCH} component={MainPageContainer} />
      <Route path={RouterPaths.ERROR} render={({ history }) => {
        return <ErrorPage navigateToHome={() => history.push(RouterPaths.HOME)} />  
      }} />
      <Redirect to={RouterPaths.ERROR}/>
    </Switch>
  )
};

export default Routs;
