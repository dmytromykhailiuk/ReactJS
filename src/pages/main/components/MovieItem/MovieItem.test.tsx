import { movies } from 'mocks/movies.mock';
import { mount, ReactWrapper } from 'enzyme';
import React from 'react';
import toJson from 'enzyme-to-json';
import { isClickInside } from 'shared/helpers';
import { history } from 'router';
import MovieItem from './MovieItem';

jest.mock('shared/helpers');

describe('MovieItem', () => {
  let wrapper: ReactWrapper<any, Readonly<{}>, React.Component<{}, {}, any>>;
  const onEditMovie = jest.fn();
  const onDeleteMovie = jest.fn();

  beforeEach(() => {
    jest.clearAllMocks();

    wrapper = mount(<MovieItem movie={movies[0]} onEditMovie={onEditMovie} onDeleteMovie={onDeleteMovie} />);
  });

  it('should match snapshot', () => {
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it("should show 'movie__menu' after click on 'movie__three-dots'", () => {
    wrapper.find('.movie__three-dots').simulate('click');
    expect(wrapper.exists('.movie__menu')).toBeTruthy();
  });

  it("should call 'onDeleteMovie'", () => {
    wrapper.find('.movie__three-dots').simulate('click');
    wrapper.find('.movie-menu__button').last().simulate('click');
    expect(onDeleteMovie).toHaveBeenCalledWith(movies[0]);
  });

  it("should call 'onEditMovie'", () => {
    wrapper.find('.movie__three-dots').simulate('click');
    wrapper.find('.movie-menu__button').first().simulate('click');
    expect(onEditMovie).toHaveBeenCalledWith(movies[0]);
  });

  it("should call 'history.push' with `/film/${movie.id}`", () => {
    (isClickInside as jest.Mock).mockImplementation(() => true);
    const push = jest.spyOn(history, 'push');
    wrapper = mount(<MovieItem movie={movies[0]} onEditMovie={onEditMovie} onDeleteMovie={onDeleteMovie} />);
    wrapper.find('.movie').simulate('click');
    expect(push).toHaveBeenCalledWith(`/film/${movies[0].id}`);
  });
});
