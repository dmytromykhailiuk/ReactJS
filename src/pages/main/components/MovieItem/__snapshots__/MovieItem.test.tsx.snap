// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MovieItem should match snapshot 1`] = `
<Memo()
  movie={
    Object {
      "budget": 55000000,
      "genres": Array [
        "Drama",
        "Romance",
      ],
      "id": 337167,
      "overview": "Believing they have left behind shadowy figures from their past, newlyweds Christian and Ana fully embrace an inextricable connection and shared life of luxury. But just as she steps into her role as Mrs. Grey and he relaxes into an unfamiliar stability, new threats could jeopardize their happy ending before it even begins.",
      "poster_path": "https://image.tmdb.org/t/p/w500/3kcEGnYBHDeqmdYf8ZRbKdfmlUy.jpg",
      "release_date": "2018-02-07",
      "revenue": 136906000,
      "runtime": 106,
      "tagline": "Don't miss the climax",
      "title": "Fifty Shades Freed",
      "vote_average": 6.1,
      "vote_count": 1195,
    }
  }
  onDeleteMovie={[MockFunction]}
  onEditMovie={[MockFunction]}
>
  <div
    className="movie"
    onClick={[Function]}
  >
    <div
      className="movie__image-wrapper"
    >
      <img
        alt="Fifty Shades Freed"
        className="movie__image"
        onError={[Function]}
        src="https://image.tmdb.org/t/p/w500/3kcEGnYBHDeqmdYf8ZRbKdfmlUy.jpg"
      />
      <div
        className="movie__three-dots"
        onClick={[Function]}
      >
        <ThreeDotsIcon>
          <div
            className="icon"
          >
            <svg
              className="bi bi-three-dots-vertical"
              fill="currentColor"
              viewBox="0 0 16 16"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M9.5 13a1.5 1.5 0 1 1-3 0 1.5 1.5 0 0 1 3 0zm0-5a1.5 1.5 0 1 1-3 0 1.5 1.5 0 0 1 3 0zm0-5a1.5 1.5 0 1 1-3 0 1.5 1.5 0 0 1 3 0z"
              />
            </svg>
          </div>
        </ThreeDotsIcon>
      </div>
    </div>
    <div
      className="movie__footer"
    >
      <div>
        <div
          className="movie__name"
        >
          Fifty Shades Freed
        </div>
        <div
          className="movie__category"
        >
          Drama, Romance
        </div>
      </div>
      <div
        className="movie__release"
      >
        2018
      </div>
    </div>
  </div>
</Memo()>
`;
